python/runtime/cudaq/algorithms/py_sample.cpp

```diff
@@ -99,7 +99,7 @@ async_sample_result pySampleAsync(kernel_builder<> &builder,
   std::hash<std::string> hasher;
   // Create a unique integer key that combines the kernel name
   // and the validated args.
-  std::size_t uniqueHash = hasher(kernelName) + hasher(py::str(args));
+  std::size_t uniqueHash = hasher(kernelName) + hasher(py::str(args)) + qpu_id;
```

Comment:
https://github.com/NVIDIA/cuda-quantum/blob/main/python/runtime/cudaq/algorithms/py_observe.cpp#L80 has the same hash formula. Is it vulnerable to the same problem?

Reply:
Yes good catch. I'll make the update.

Reply:
It may be helpful to gracefully avoid collisions. I tested the following gist which avoids collisions and it appears to work (even without the `+ qpu_id` change), and I believe it would be insensitive to collisions.
https://gist.github.com/bmhowe23/429773e97aea6d2a3c21b8650b0bd178

(The corresponding changes for `observe_async` may need adjustments, though. `observe_async` seems to be failing for me even with the `+ qpu_id` change, so I can't properly test out this change there.)

Reply:
Playing devil's advocate here :-) 

What do you think about removing the `asyncArgsHolder` cache map altogether and instead attaching the `OpaqueArguments` to the functor itself?

e.g.,
```
  // Add the opaque args to the holder and pack the args into it
  auto argsHolder = std::make_unique<OpaqueArguments>();
  packArgs(*argsHolder, validatedArgs);
  return details::runSamplingAsync(
      detail::make_copyable_function(
          [&builder, argData = std::move(argsHolder)]() mutable {
            builder.jitAndInvoke(argData->data());
          }),
      platform, kernelName, shots, qpu_id);
```

Reply:
Thanks to both of you for the suggestions. Thien I like your approach and have updated the code with it for both sample and observe